x-pack/solutions/security/plugins/security_solution/server/utils/open_api_tool/open_api_tool.ts

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type Oas from 'oas';
+import type { JsonSchema } from '@n8n/json-schema-to-zod';
+import { jsonSchemaToZod } from '@n8n/json-schema-to-zod';
+import type { SchemaObject } from 'oas/dist/types.cjs';
+import { z } from '@kbn/zod';
+import type { StructuredToolInterface } from '@langchain/core/tools';
+import { groupBy, memoize } from 'lodash';
+import { fixOpenApiSpecIteratively, formatToolName, isOperation } from './utils';
+import type { Operation } from './utils';
+
+export abstract class OpenApiTool<T> {
+  private dereferencedOas: Oas;
+  private getParametersAsZodSchemaMemoized = memoize(
+    this.getParametersAsZodSchemaInternal.bind(this),
+    (args) => {
+      return args.operation.getOperationId();
+    }
+  );
+
+  protected constructor(args: { dereferencedOas: Oas }) {
+    const { dereferencedOas } = args;
+    this.dereferencedOas = dereferencedOas;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  static fixOpenApiSpecIteratively(openApiSpec: any): any {
+    return fixOpenApiSpecIteratively(openApiSpec);
+  }
+
+  protected getOperations() {
+    const oas = this.dereferencedOas;
+    return Object.values(oas.getPaths())
+      .map((operationByMethod) => Object.values(operationByMethod))
+      .flat()
```

Comment:
nit to use `flatMap` here
```suggestion
      .flatMap(Object.values)
```

Also can you add this in tests?